var/spack/repos/builtin/packages/mvapich2-gdr/package.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import sys
+from spack import *
+
+class Mvapich2Gdr(AutotoolsPackage):
+    # Mvapich2-Gdr is not installable from source and is only available through the binary mirror below
+    # http://mvapich.cse.ohio-state.edu:8080/download/mvapich/spack-mirror/mvapich2-gdr/
+
+    homepage = "http://mvapich.cse.ohio-state.edu"
+    url      = "http://mvapich.cse.ohio-state.edu:8080/download/mvapich/spack-mirror/mvapich2-gdr/mvapich2-gdr-2.3.4.tar.gz"
+    
+    version('2.3.4', sha256='ed78101e6bb807e979213006ee5f20ff466369b01f96b6d1cf0c471baf7e35aa')
+    version('2.3.3', sha256='9b7b5dd235dbf85099fba3b6f1ccb49bb755923efed66ddc335921f44cb1b8a8')
+    
+    provides('mpi')
+    provides('mpi@:3.1')
+
+    variant(
+        'cuda_version',
+        description='Currently supported for CUDA versions 9.2, 10.1 and 10.2',
+        default='10.1',
+        values=('10.2', '10.1', '9.2'),
+        multi=False
+    )
+
+    variant(
+        'process_managers',
+        description='The process manager to activate.',
+        default='mpirun',
+        values=('slurm', 'mpirun', 'pbs', 'jsrun'),
+        multi=False
+    )
+    
+    variant(
+        'distribution',
+        description='The type of fabric distribution.',
+        default='stock-ofed',
+        values=('stock-ofed', 'mofed4.5', 'mofed4.6', 'mofed4.7', 'mofed5.0'),
+        multi=False
+    )
+     
+    variant(
+        'pmi_version',
+        description='The pmi version to be used with slurm.' \
+                    'Is ignored if set for mpirun or jsrun.' \
+                    'jsrun uses pmix regardless of chosen option.',
+        default='pmi1',
+        values=('pmi1', 'pmi2', 'pmix'),
+        multi=False
+    )
+
+    variant(
+        'mcast',
+        description='Enable/Disable support for mcast',
+        default=True
+    )
+
+    variant(
+        'openacc',
+        description='Enable/Disable support for openacc',
+        default=False
+    )
+        
+    variant(
+        'core_direct',
+        description='Enable/Disable support for core_direct',
+        default=False
+    )
+
+    depends_on('bison@3.4.2')
+    depends_on('libpciaccess@0.13.5', when=(sys.platform != 'darwin'))
+    depends_on('libxml2@2.9.10')
+    depends_on('cuda@9.2.88', when='cuda_version=9.2')
+    depends_on('cuda@10.1.243', when='cuda_version=10.1')
+    depends_on('cuda@10.2.89', when='cuda_version=10.2')
+    depends_on('pmix@3.1.3', when='pmi_version=pmix')
```

Comment:
Do all of these dependencies need to be these exact versions, or is that the lowest acceptable version? If the latter, add a colon to the end to create an open-ended version range.

Reply:
If the maintainers of the dependency update their package, (eg. bison updated theirs recently), the hash value of MVAPICH2-X and MVAPICH2-GDR binaries would become outdated and `$spack install` would fail to find them on the binary mirror unless the user specifies the exact version of the dependency that the binaries were created with. 

This is because the hashes would now be generated based on the new version of the dependencies.

Reply:
Yes, this is true for every single package in Spack. But if every package chooses a random version to link to, you won't be able to have a single DAG with a single version of a library.

There is ongoing work on a new concretizer that will consider already available binary versions before choosing which version to install. Until then, you should allow the dependency to choose as wide of a range of versions as possible.

Reply:
Use `spack buildcache list -lv` to list the buildcaches available including the variants used and the hash.
Use `spack buildcache install /hash` to install the package and all of it's link and run dependencies. 
The hash is shorthand for the full spec.

Reply:
Thanks, I understand the reasoning behind your suggestions. But, I did face the following issue which required the addition of explicit versions of dependencies.
 
When we created our binaries, they were dependent on the latest versionof bison, at that time - 3.4.2 `depends_on('bison')`. However, after a certain date, the `spack install` command stopped working for all binaries in our mirror. And we found that this was due to the update of the bison package which now has 3.6.4 as its latest version.

So, unless users specify `^bison@3.4.2`, spack install will not find any of the binaries in our mirror. How would we resolve this without creating new binaries for each version of bison?

Reply:
Hmm, I would add binaries for the latest version of bison, that's what Spack will install by default. Once the new concretizer comes out (next Spack release) it will be smart enough to use the binary version even if it's older.

Reply:
Given that we will be publicly announcing the availability of MVAPICH2-GDR and MVAPICH2-X on spack to all our users, I believe it would be better if currently, we have the explicit versions so that we're not affected if a dependency updates their package. Once the new release of spack is out, I can remove version specifications for the dependencies and regenerate the binaries and raise a new pull request. Would that be feasible?